R/derive_vars_cat.R

```diff
@@ -0,0 +1,218 @@
+#' In this case, the middle condition should be:
+#' `AVAL <= 170 & AVAL > 160`
+#'
+#' @return data frame
+#' @family der_gen
+#' @keywords der_gen
+#' @export
+#'
+#' @examples
+#'
+#' advs <- tibble::tribble(
+#'   ~USUBJID, ~VSTEST, ~AVAL,
+#'   "01-701-1015", "Height", 147.32,
+#'   "01-701-1015", "Weight", 53.98,
+#'   "01-701-1023", "Height", 162.56,
+#'   "01-701-1023", "Weight", 78.47,
+#'   "01-701-1028", "Height", 177.8,
+#'   "01-701-1028", "Weight", 98.88,
+#'   "01-701-1033", "Height", 175.26,
+#'   "01-701-1033", "Weight", 88.45,
+#'   "01-701-1034", "Height", 154.94,
+#'   "01-701-1034", "Weight", 63.5,
+#'   "01-701-1047", "Height", 148.59,
+#'   "01-701-1047", "Weight", 66.23,
+#'   "01-701-1097", "Height", 168.91,
+#'   "01-701-1097", "Weight", 78.02,
+#'   "01-701-1111", "Height", 158.24,
+#'   "01-701-1111", "Weight", 60.33,
+#'   "01-701-1115", "Height", 181.61,
+#'   "01-701-1115", "Weight", 78.7,
+#'   "01-701-1118", "Height", 180.34,
+#'   "01-701-1118", "Weight", 71.67
+#' )
+#'
+#' definition <- exprs(
+#'   ~condition, ~AVALCAT1, ~AVALCA1N, ~NEWCOL,
+#'   VSTEST == "Height" & AVAL > 160, ">160 cm", 1, "extra1",
+#'   VSTEST == "Height" & AVAL <= 160, "<=160 cm", 2, "extra2"
+#' )
+#' derive_vars_cat(
+#'   dataset = advs %>% dplyr::filter(VSTEST == "Height"),
+#'   definition = definition
+#' )
+#' # using by_vars:
+#' definition2 <- exprs(
+#'   ~VSTEST, ~condition, ~AVALCAT1, ~AVALCA1N,
+#'   "Height", AVAL > 160, ">160 cm", 1,
+#'   "Height", AVAL <= 160, "<=160 cm", 2,
+#'   "Weight", AVAL > 70, ">70 kg", 1,
+#'   "Weight", AVAL <= 70, "<=70 kg", 2
+#' )
+#'
+#' derive_vars_cat(
+#'   dataset = advs,
+#'   definition = definition2,
+#'   by_vars = exprs(VSTEST)
+#' )
+#'
+#' # With three conditions:
+#' definition3 <- exprs(
+#'   ~VSTEST, ~condition, ~AVALCAT1, ~AVALCA1N,
+#'   "Height", AVAL > 170, ">170 cm", 1,
+#'   "Height", AVAL <= 170 & AVAL > 160, "<=170 cm", 2,
+#'   "Height", AVAL <= 160, "<=160 cm", 3
+#' )
+#'
+#' derive_vars_cat(
+#'   dataset = advs,
+#'   definition = definition3,
+#'   by_vars = exprs(VSTEST)
+#' )
+derive_vars_cat <- function(dataset,
+                            definition,
+                            by_vars = NULL) {
+  # assertions
+  assert_data_frame(dataset)
+  assert_expr_list(definition)
+  if (!is.null(by_vars)) {
+    assert_vars(by_vars)
+    assert_data_frame(dataset,
+      required_vars = c(
+        admiraldev::extract_vars(definition) %>% unique(),
+        by_vars
+      )
+    )
+  } else {
+    assert_data_frame(dataset, required_vars = admiraldev::extract_vars(definition) %>% unique())
+  }
+
+  # transform definition to tibble
+  names(definition) <- NULL
+  definition <- tryCatch(
+    {
+      tibble::tribble(!!!definition)
+    },
+    error = function(e) {
+      # Catch the error and append your own message
+      stop("Failed to convert `definition` to tribble: ", e$message)
```

Comment:
See below message in snapshot test please

Reply:
I made the error message more verbose.